### DISCUSSION TITLE
Guide to migrate ?

### DISCUSSION THREAD

**Discussion by noahfraiture** [Aylur/ags#615] [Ideas]
Hello,
It would be great to have a small documentation to how to make the migration to ags v2. For instance I have https://github.com/noahfraiture/nix/tree/main/modules/home-manager/ags and I read that `Service` and `Utils` was not available anymore. Is there lot of change to do ?

**Comment by dtomvan** [comment]
@noahfraiture Yes, you basically need to start from scratch:
 https://github.com/Aylur/ags/issues/618#issuecomment-2476866842

How to do that: https://aylur.github.io/ags/guide/quick-start.html

Edit: as I'm appearantely the first comment in the thread I'll link to Aylur's migration guide as well: https://aylur.github.io/ags/guide/migrate.html

Edit2: the Astal repo has an example if you quickly want a simple bar using ags v2: https://github.com/Aylur/astal/tree/main/examples/js/simple-bar

**Reply by noahfraiture** [reply]
Oh. Ok thank you

**Comment by dwilliam62** [comment]
Rather disappointed how this was managed.   Having to redo everything from scratch is not an easy task.

**Reply by mylinuxforwork** [reply]
> Rather disappointed how this was managed. Having to redo everything from scratch is not an easy task.

@dwilliam62 Agree. Now, I need to remove the ags sidebar of my public dotfiles for now or switch back to eww. I can imagine that v2 is more powerful and structured. But developers need time to learn.

**Reply by dwilliam62** [reply]
yeah.   There is this  https://github.com/kotontrion/PKGBUILDS/blob/main/agsv1/PKGBUILD    I tried to build it for another project that also uses AGS.  It built and installed but didn't work for me.   The executable is agsv1 but even linking it to ags didn't help.   You might have better luck  I too really wished v2  had been a new package and just depreciate the old one for compatibility.   Some projects used AGS for the waybar and more.

**Reply by Aylur** [reply]
AGS v1 has been pretty much abandoned for months now. I've added a notice sometime ago both on the front page of the wiki and the readme on github. I went on to push for AGSv2 instead of just abandoning the project because
- I just did not want to archive a +2000 star repo
- I wanted to make it clear that v1 is no longer maintained
- It forces adaption

You still have the option of not updating, you don't "have to" start from scratch immediately.
The AUR maintainer provided a pkgbuild for you. On NixOS you can pin the commit of the flake.
I'm not in touch with other package maintainers you will have to refer to them or your package manager on how to downgrade.

I understand that it might be annoying that updating your system broke your current setup, but this is on you. It will happen sometimes when you are on a rolling distribution where you don't pin the version of software that you use.

**Reply by dwilliam62** [reply]
It's more than 'annoying'   mylinuxforwork  and others create hyprland environments to help lower the bar for people starting out with Linux or Hyprland.  Other projects have built complete  waybar like applications, menus, etc, using AGS.  Users upgrade the OS  and all that breaks. Newer users don't know how to pin or build from source.   I've spent today testing how to make AGS v1 work for Arch, Fedora, NIXOS.   So it can be documented and limit the impact of this change.

**Reply by mylinuxforwork** [reply]
Great. Can you send me a feedback? Will check today ags v2. Let’s see how complicated it is to migrate the sidebar and calendar to the new version.

**Reply by kurtbahartr** [reply]
> It's more than 'annoying' mylinuxforwork and others create hyprland environments to help lower the bar for people starting out with Linux or Hyprland. Other projects have built complete waybar like applications, menus, etc, using AGS. Users upgrade the OS and all that breaks. Newer users don't know how to pin or build from source. I've spent today testing how to make AGS v1 work for Arch, Fedora, NIXOS. So it can be documented and limit the impact of this change.

@dwilliam62, I did the hard work for you if you still couldn't figure it out on Arch. You can check the diff over at https://github.com/dianaw353/dotfiles/compare/V1.1.5...V1.1.6 to see how I handled it.

I also did little start on AGSv2 migration with https://github.com/dianaw353/dotfiles/compare/V1.1.6...V1.1.7, though that's just the beginning based on the migration guide and still needs lots of work or even a complete start from scratch for that matter.

**Reply by kurtbahartr** [reply]
> Great. Can you send me a feedback? Will check today ags v2. Let’s see how complicated it is to migrate the sidebar and calendar to the new version.

@mylinuxforwork, I tried to port over the entry point (`config.ts`) to AGSv2 last night until 2 AM (ADHD+insomnia gaming) but I rage quit as it would always complain that something wasn't right.

I'm sure there's a way to fix those errors - I just couldn't figure out how back then while being half-asleep.

**Reply by dwilliam62** [reply]
> 
> 
> @dwilliam62, I did the hard work for you if you still couldn't figure it out on Arch. You can check the diff over at [dianaw353/dotfiles@V1.1.5...V1.1.6](https://github.com/dianaw353/dotfiles/compare/V1.1.5...V1.1.6) to see how I handled it.
> 
> I also did little start on AGSv2 migration with [dianaw353/dotfiles@V1.1.6...V1.1.7](https://github.com/dianaw353/dotfiles/compare/V1.1.6...V1.1.7), though that's just the beginning based on the migration guide and still needs lots of work or even a complete start from scratch for that matter.

  Thanks!  I passed it on the main dev.  See what he plans to do long term about it.  We have a working solution for now. If he decides to stay with AGS will look to leverage your solution so they can both be installed. Making migration easier. 
  
  For NIX I just created an override I will check out your flake later.   Thank you again for all the work

**Comment by mylinuxforwork** [comment]
Is there a way to offer a seperate AUR or general installation instruction for ags v1? And let the user decide when to do the switch?

**Reply by kurtbahartr** [reply]
```sh
mkdir -p ~/agsv1-rebuild
cd ~/agsv1-rebuild
wget https://github.com/kotontrion/PKGBUILDS/raw/main/agsv1/PKGBUILD # or curl it
makepkg -siC
```

If you want an Ansible task to handle it all instead: https://github.com/dianaw353/dotfiles/compare/V1.1.5...V1.1.6#diff-1969cc5c0f44cf596c8102df452d2407f1f1bd404fe17f922e4b69d54a201a50R47-R82

Of course you'll have to change every single part of your code that calls the `ags` executable to call `agsv1` instead. The files and the syntax may stay as is - It's just the executable name that changed so that both versions can coexist.

I also added a handy keybind accompanied by a script in Diana's dotfiles should you need to install both versions and have the ability to switch between both;
Keybind (`SUPER+Shift+S`, or what I'd call a "Triple-S"): https://github.com/dianaw353/dotfiles/compare/V1.1.6...V1.1.7#diff-5e00db9a2dde713e5f035ad06dc01fe2d311eef94d439a59a0ccd82e64f1bc8cR33
Script: https://github.com/dianaw353/dotfiles/blob/V1.1.7/roles/hyprland/files/hypr/scripts/switch-ags.sh

And if you want to use the exact same keybinds on both AGS versions without modifying keybinds by hand, you might wanna write a script like this;
```sh
#!/bin/bash
pgrep ags > /dev/null
if [[ "$?" != "0" ]]; then
  agsv1 -t launcher
else
  ags toggle launcher
fi
```
> For `--run-js` and `request`, replace `agsv1 -t $widget` with `agsv1 -r "$js_syntax"` and `ags toggle $widget` with `ags request $request`.

Then replace `agsv1` commands with those scripts in your keybind configs.

**Comment by Aylur** [comment]
You can not avoid going through the new [Astal wiki](https://aylur.github.io/astal/guide/typescript/first-widgets) in order to get a grasp on how things work.
I've written however a [migration guide](https://aylur.github.io/ags/guide/migrate.html) that highlights what changes you will have to make.
Let me know if there is anything you feel is missing

**Reply by mylinuxforwork** [reply]
Thanks for providing the migration guide. Still a lot to learn but it is a super starting point.

**Reply by kurtbahartr** [reply]
Thanks for the migration guide! I'm sure it will come in handy for @dianaw353 and me throughout the coming weekend! <3

**Comment by dtomvan** [comment]
For Nix users I've written a nix flake, similar to the Arch Linux [PKGBUILD](https://github.com/kotontrion/PKGBUILDS/blob/main/agsv1/PKGBUILD), so you can keep both versions of ags side-by-side (while working on a new config). Includes a quick-and-dirty home-manager module that points agsv1 to a different directory than `~/.config/ags`. Everything you need to know is in the repo: https://github.com/dtomvan/agsv1

**Comment by JaKooLit** [comment]
My request is not to take down or remove the v1.8.2 from the releases :)

since Im pulling from that

**Reply by kurtbahartr** [reply]
It has already been superseded by v2 in the AUR.

**Reply by dwilliam62** [reply]
Yup he understands.   He means the tar ball or PKGBUILD file.   Since for the time being he has to stay on v1 for his AGS app.  He can't port it to v2 at this time.

**Comment by dianaw353** [comment]
If anyone needs I have created a script that quickly switches between agsv1 and agsv2 located [here](https://github.com/dianaw353/dotfiles/blob/main/roles/hyprland/files/hypr/scripts/switch-ags.sh). I believe this would be helpful for those who are trying to develop for v2 but don't want to have the hassle of switching back and forth between v2 and v1

The only requirement for the script to work I believe is the following file locations:
agsv1: `~/.config/ags`
agsv2: `~/.config/agsv2`

**Reply by mylinuxforwork** [reply]
Great. Thank you.
